File: packages/sdk/src/resources/usage.ts
/**
 * Usage-based billing resource client — metering + prepaid credits.
 *
 * Maps to `/v1/usage` (record / meter / charge / limit / credits).
 */
import type { UsageMeter, CreditBalance, Money } from "@settlekit/common";
import type { HttpClient, RequestOptions } from "../http-client.js";

/** Identity of a metered dimension. */
export interface MeterRef {
  organizationId: string;
  customerId: string;
  productId: string;
  metric: string;
}

/** Identity of a prepaid balance. */
export interface BalanceRef {
  organizationId: string;
  customerId: string;
  productId: string;
}

/** Result of a usage-limit check. */
export interface LimitCheck {
  withinLimit: boolean;
  used: number;
  limit: number;
  remaining: number;
}

/** Client for usage-based billing endpoints. */
export class UsageResource {
  constructor(private readonly http: HttpClient) {}

  /** Record `quantity` units of a metric (creates the meter on first use). */
  record(input: MeterRef & { quantity?: number; periodStart?: string }, options?: RequestOptions): Promise<UsageMeter> {
    return this.http.post<UsageMeter>("/v1/usage/record", input, options);
  }

  /** Compute the charge for a meter's usage at a unit price. */
  charge(input: MeterRef & { unitAmount: string; periodStart?: string }, options?: RequestOptions): Promise<Money> {
    return this.http.post<Money>("/v1/usage/charge", input, options);
  }

  /** Evaluate usage against a hard limit. */
  limit(input: MeterRef & { limit: number; periodStart?: string }, options?: RequestOptions): Promise<LimitCheck> {
    return this.http.post<LimitCheck>("/v1/usage/limit", input, options);
  }

  /** Read a prepaid credit balance. */
  credits(ref: BalanceRef, options?: RequestOptions): Promise<CreditBalance | null> {
    return this.http.get<CreditBalance | null>(
      `/v1/usage/credits?organizationId=${encodeURIComponent(ref.organizationId)}&customerId=${encodeURIComponent(ref.customerId)}&productId=${encodeURIComponent(ref.productId)}`,
      options,
    );
  }

  /** Grant prepaid credits. */
  grantCredits(input: BalanceRef & { credits: number }, options?: RequestOptions): Promise<CreditBalance> {
    return this.http.post<CreditBalance>("/v1/usage/credits/grant", input, options);
  }

  /** Consume prepaid credits (meter a paid call). */
  consumeCredits(input: BalanceRef & { credits?: number }, options?: RequestOptions): Promise<CreditBalance> {
    return this.http.post<CreditBalance>("/v1/usage/credits/consume", input, options);
  }
}

File: sdks/go/usage.go
package settlekit

import (
	"context"
	"net/http"
)

// RecordUsageInput is the body for RecordUsage (POST /v1/usage/record).
type RecordUsageInput struct {
	OrganizationID string         `json:"organizationId"`
	CustomerID     string         `json:"customerId"`
	ProductID      string         `json:"productId"`
	Metric         string         `json:"metric"`
	Quantity       int            `json:"quantity"`
	Metadata       map[string]any `json:"metadata,omitempty"`
}

// RecordUsage records a metered-usage event. POST /v1/usage/record.
func (c *Client) RecordUsage(ctx context.Context, in RecordUsageInput) (*UsageRecord, error) {
	var out UsageRecord
	if err := c.do(ctx, http.MethodPost, "/v1/usage/record", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetCredits returns an organization's prepaid credit balance.
// GET /v1/usage/credits[?organizationId=].
func (c *Client) GetCredits(ctx context.Context, organizationID string) (*CreditBalance, error) {
	q := newQuery()
	q.add("organizationId", organizationID)
	var out CreditBalance
	if err := c.do(ctx, http.MethodGet, "/v1/usage/credits"+q.encode(), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GrantCredits adds prepaid credits to an organization's balance.
// POST /v1/usage/credits/grant.
func (c *Client) GrantCredits(ctx context.Context, organizationID string, credits int) (*CreditBalance, error) {
	body := map[string]any{"organizationId": organizationID, "credits": credits}
	var out CreditBalance
	if err := c.do(ctx, http.MethodPost, "/v1/usage/credits/grant", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ConsumeCredits debits prepaid credits from an organization's balance.
// POST /v1/usage/credits/consume.
func (c *Client) ConsumeCredits(ctx context.Context, organizationID string, credits int) (*CreditBalance, error) {
	body := map[string]any{"organizationId": organizationID, "credits": credits}
	var out CreditBalance
	if err := c.do(ctx, http.MethodPost, "/v1/usage/credits/consume", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
